optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintSessionFactory.java

```diff
@@ -45,6 +47,7 @@ public DroolsConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDesc
     @Override
     public ConstraintSession<Solution_> buildSession(boolean constraintMatchEnabled, Solution_ workingSolution) {
         KieSession kieSession = kieBase.newKieSession();
+        ((RuleEventManager) kieSession).addEventListener(new OptaPlannerRuleEventListener()); // Enables undo in rules
```

Comment:
@ge0ffrey I don't really like the fact that I'm referencing an implementation class from a different unrelated package. That said, I couldn't find a better place for it - suggestions?

Reply:
(At the very least, I think it should be together with the actual undo listener. But that is in the public API, so I can not leave this listener there.)

Reply:
This is a good start, so I wouldn't stop the merge, but I agree.
Both OptaPlannerRuleEventListener and GLOBAL_SCORE_HOLDER_KEY (in DroolsScoreDirector) are used by both drl and CS impl. We'll probably want to move them into for example the package org.optaplanner.core.impl.score.drools ? Definitively impl indeed.

Reply:
ConstraintActivationUnMatchListener can also go into org.optaplanner.core.impl.score.drools, in 8.0. Add it to https://issues.jboss.org/browse/PLANNER-750 so we don't forget to do that.